### PR TITLE
data-json - Add conditional template

### PR DESCRIPTION
**Pull request by duboisp**
Add the possibility to do conditional template with the data-json plugin.

Test to do at least:
* [x] Ensure the data-json working example work the same as before
* [x] Ensure the data-json HTML5 template working example work the same as before
* [x] Ensure the JSON manager working example work the same as before
* [x] Test the functional aspect of the conditional template via its working example
* [x] Review the code.

Working example: https://universallabs.org/wet/labs/conditional-tmpl/unmin/demos/wb-data-json/conditional-en.html

### CODE REVIEW

**Comment by duboisp** [comment]
@GormFrank for discussion to know who would be able to review this PR.

**Comment by duboisp** [comment]
Pre-approved upon successful review and local testing by @polmih

File: site/layouts/assessment_wrote_en-en.hbs [hunk 1]
```diff
@@ -170,6 +170,111 @@
...
+					{
+						"test": "fn:isLiteral",
+						"assess": "/acr:asset",
+						"mapping": [
+							{
+								"@type": "rdf:Alt",
+								"mapping": [
+									{
+										"template": "[data-asset-literal]",
+										"test": "fn:guestType",
```
Comment (polmih) [inline site/layouts/assessment_wrote_en-en.hbs:186]:
```suggestion
										"test": "fn:guessType",
```

File: site/layouts/assessment_wrote_en-en.hbs [hunk 2]
```diff
@@ -170,6 +170,111 @@
...
+						"test": "fn:guestType",
+						"value": "/acr:asset",
+						"expect": "xsd:anyURI",
+						"mapping": [
+							{ "selector": "a", "value": "/acr:asset", "attr": "href" }
+						]
+					},
+					{
+						"template": "[data-asset-code]",
+						"test": "fn:guestType",
```
Comment (polmih) [inline site/layouts/assessment_wrote_en-en.hbs:229]:
```suggestion
						"test": "fn:guessType",
```

File: site/layouts/assessment_wrote_en-en.hbs [hunk 3]
```diff
@@ -170,6 +170,111 @@
...
+									"value": "/",
+									"expect": "xsd:anyURI",
+									"append": true,
+									"mapping": [
+										{ "selector": "a", "value": "/@value", "attr": "href" }
+									]
+								},
+								{
+									"template": "[data-source-code]",
+									"test": "fn:guestType",
```
Comment (polmih) [inline site/layouts/assessment_wrote_en-en.hbs:254]:
```suggestion
									"test": "fn:guessType",
```

File: site/layouts/assessment_wrote_en-en.hbs [hunk 4]
```diff
@@ -170,6 +170,111 @@
...
+									"value": "/",
+									"expect": "xsd:string",
+									"append": true,
+									"mapping": [
+										{ "selector": "code", "value": "/@value" }
+									]
+								},
+								{
+									"template": "[data-asset-image]",
+									"test": "fn:guestType",
```
Comment (polmih) [inline site/layouts/assessment_wrote_en-en.hbs:264]:
```suggestion
									"test": "fn:guessType",
```

File: site/layouts/assessment_wrote_en-fr.hbs [hunk 5]
```diff
@@ -169,6 +169,111 @@
...
+					{
+						"test": "fn:isLiteral",
+						"assess": "/acr:asset",
+						"mapping": [
+							{
+								"@type": "rdf:Alt",
+								"mapping": [
+									{
+										"template": "[data-asset-literal]",
+										"test": "fn:guestType",
```
Comment (polmih) [inline site/layouts/assessment_wrote_en-fr.hbs:185]:
```suggestion
										"test": "fn:guessType",
```

File: site/layouts/assessment_wrote_en-fr.hbs [hunk 6]
```diff
@@ -169,6 +169,111 @@
...
+										"value": "/acr:asset",
+										"expect": "xsd:anyURI",
+										"mapping": [
+											{ "selector": "a", "value": "/acr:asset", "attr": "href" }
+										]
+									},
+									{
+										"template": "[data-asset-code]",
+										"test": "fn:guestType",
```
Comment (polmih) [inline site/layouts/assessment_wrote_en-fr.hbs:194]:
```suggestion
										"test": "fn:guessType",
```

File: site/layouts/assessment_wrote_en-fr.hbs [hunk 7]
```diff
@@ -169,6 +169,111 @@
...
+											{ "selector": "code", "value": "/acr:asset" }
+										]
+									}
+								]
+							}
+						]
+					},
+					{
+						"template": "[data-asset-image]",
+						"test": "fn:guestType",
```
Comment (polmih) [inline site/layouts/assessment_wrote_en-fr.hbs:207]:
```suggestion
						"test": "fn:guessType",
```

File: site/layouts/assessment_wrote_en-fr.hbs [hunk 8]
```diff
@@ -169,6 +169,111 @@
...
+						"mapping": [
+							{ "selector": "img", "value": "/acr:asset/acr:content/@value", "attr": "src" },
+							{ "selector": "a", "value": "/acr:asset/acr:content/@value", "attr": "href" },
+							{ "selector": "img", "value": "/acr:asset/dct:title", "attr": "alt" }
+
+						]
+					},
+					{
+						"template": "[data-asset-literal]",
+						"test": "fn:guestType",
```
Comment (polmih) [inline site/layouts/assessment_wrote_en-fr.hbs:219]:
```suggestion
						"test": "fn:guessType",
```

File: site/layouts/assessment_wrote_en-fr.hbs [hunk 9]
```diff
@@ -169,6 +169,111 @@
...
+						"value": "/acr:asset",
+						"expect": "xsd:anyURI",
+						"mapping": [
+							{ "selector": "a", "value": "/acr:asset", "attr": "href" }
+						]
+					},
+					{
+						"template": "[data-asset-code]",
+						"test": "fn:guestType",
```
Comment (polmih) [inline site/layouts/assessment_wrote_en-fr.hbs:228]:
```suggestion
						"test": "fn:guessType",
```

File: site/layouts/assessment_wrote_en-fr.hbs [hunk 10]
```diff
@@ -169,6 +169,111 @@
...
+									"value": "/",
+									"expect": "xsd:anyURI",
+									"append": true,
+									"mapping": [
+										{ "selector": "a", "value": "/@value", "attr": "href" }
+									]
+								},
+								{
+									"template": "[data-source-code]",
+									"test": "fn:guestType",
```
Comment (polmih) [inline site/layouts/assessment_wrote_en-fr.hbs:253]:
```suggestion
									"test": "fn:guessType",
```

File: site/layouts/assessment_wrote_en-fr.hbs [hunk 11]
```diff
@@ -169,6 +169,111 @@
...
+									"value": "/",
+									"expect": "xsd:string",
+									"append": true,
+									"mapping": [
+										{ "selector": "code", "value": "/@value" }
+									]
+								},
+								{
+									"template": "[data-asset-image]",
+									"test": "fn:guestType",
```
Comment (polmih) [inline site/layouts/assessment_wrote_en-fr.hbs:263]:
```suggestion
									"test": "fn:guessType",
```

File: src/plugins/wb-data-json/conditional-en.hbs [hunk 12]
```diff
@@ -0,0 +1,1488 @@
...
+
+
+<p>The following list is an example of a conditional mapping by guessing the data type.</p>
+
+<ul data-wb-json='{
+	"url": "demo/conditional-1.json#/subject",
+	"mapping": [
+		{
+			"template": "[data-text]",
+			"test": "fn:guestType",
```
Comment (polmih) [inline src/plugins/wb-data-json/conditional-en.hbs:32]:
```suggestion
			"test": "fn:guessType",
```

File: src/plugins/wb-data-json/conditional-en.hbs [hunk 13]
```diff
@@ -0,0 +1,1488 @@
...
+					"template": "[data-gettype]",
+					"test": "fn:getType",
+					"assess": "/zeta",
+					"expect": "boolean",
+					"mapping": [
+						{ "selector": "span", "value": "/text/zeta" }
+					]
+				},
+				{
+					"template": "[data-guesttype]",
```
Comment (polmih) [inline src/plugins/wb-data-json/conditional-en.hbs:309]:
```suggestion
					"template": "[data-guesstype]",
```

File: src/plugins/wb-data-json/conditional-en.hbs [hunk 14]
```diff
@@ -0,0 +1,1488 @@
...
+					"test": "fn:guestType",
+					"assess": "/alpha",
+					"expect": "rdfs:Container",
+					"mapping": [
+						{ "selector": "span", "value": "/alpha/0" }
+					]
+				},
+				{
+					"template": "[data-guesttype]",
```
Comment (polmih) [inline src/plugins/wb-data-json/conditional-en.hbs:318]:
```suggestion
					"template": "[data-guesstype]",
```

File: src/plugins/wb-data-json/conditional-en.hbs [hunk 15]
```diff
@@ -0,0 +1,1488 @@
...
+					"test": "fn:guestType",
```
Comment (polmih) [inline src/plugins/wb-data-json/conditional-en.hbs:319]:
```suggestion
					"test": "fn:guessType",
```

File: src/plugins/wb-data-json/conditional-en.hbs [hunk 16]
```diff
@@ -0,0 +1,1488 @@
...
+					"assess": "/beta",
+					"expect": "rdfs:Literal",
+					"mapping": [
+						{ "selector": "span", "value": "/beta" }
+					]
+				},
+				{
+					"template": "[data-guesttype]",
```
Comment (polmih) [inline src/plugins/wb-data-json/conditional-en.hbs:327]:
```suggestion
					"template": "[data-guesstype]",
```

File: src/plugins/wb-data-json/conditional-en.hbs [hunk 17]
```diff
@@ -0,0 +1,1488 @@
...
+					"test": "fn:guestType",
```
Comment (polmih) [inline src/plugins/wb-data-json/conditional-en.hbs:328]:
```suggestion
					"test": "fn:guessType",
```

File: src/plugins/wb-data-json/conditional-en.hbs [hunk 18]
```diff
@@ -0,0 +1,1488 @@
...
+					"assess": "/gamma",
+					"expect": "rdfs:Literal",
+					"mapping": [
+						{ "selector": "span", "value": "/gamma" }
+					]
+				},
+				{
+					"template": "[data-guesttype]",
```
Comment (polmih) [inline src/plugins/wb-data-json/conditional-en.hbs:336]:
```suggestion
					"template": "[data-guesstype]",
```

Comment (polmih) [reply src/plugins/wb-data-json/conditional-en.hbs:336]: guestType sts all over this file.  I will leave it to you to do a find and replace in the code :)

**Review by polmih — DISMISSED** [review]
All plugins mentioned in the PR description were tested locally. Did not encounter any issues.

**Comment by duboisp** [comment]
Approved once the update is push and a second review / testing is completed.

**Comment by duboisp** [comment]
@GormFrank can you perform the second review? The requested change was applied and fixed the guessType behaviour when its undefined or when its point to a non-typed JSON object.

File: site/pages/docs/ref/wb-data-json/wb-data-json-en.hbs [hunk 19]
```diff
@@ -604,6 +659,12 @@
 		<td><code>data-wb-json='{ "url": "", "source": "#idToMyTemplate" }'</code></td>
 		<td>JQuery selector that represent the template element on the current page.</td>
 	</tr>
+	<tr>
+		<td><code>streamline</code></td>
+		<td>Template only. Indicate to not initiate the process by iterating the data and instead execute the mapping object directly</td>
```
Comment (GormFrank) [inline site/pages/docs/ref/wb-data-json/wb-data-json-en.hbs:664]: Slight rephrase:
```suggestion
		<td>Template only. Indicate not to initiate the process of iterating the data and instead execute the mapping object directly</td>
```

File: site/pages/docs/ref/wb-data-json/wb-data-json-en.hbs [hunk 20]
```diff
@@ -267,7 +269,7 @@
...
+<p>Since WET 4.0.27 or without the <code>streamline</code> flag since WET 4.0.64 it is possible to iterate anything that is not an array to render content through a <code>&lt;template&gt;</code>. It simply get transformed into a array. There is a low risk of lost of information, only one specific use case. It's quite similar to the JSON-LD Expanded Form. In this transformation process, it assume the array key is an IRI. Then the will result include some property from JSON-LD such as <code>@id</code> and <code>@value</code>. The <code>@id</code> wont be overwritten, in those such case the array key will be discard after the transformation.</p>
```
Comment (GormFrank) [inline site/pages/docs/ref/wb-data-json/wb-data-json-en.hbs:272]: May be worth adding a link to the IRI definition for general audience to get the reference :)

**Review by GormFrank — DISMISSED** [review]
Approved with some content changes that can be tackled after (please consider French pages as well when looking at my comments)

File: site/pages/docs/ref/wb-data-json/wb-data-json-en.hbs [hunk 21]
```diff
@@ -423,6 +425,59 @@
...
+	<dt><code>fn:getType</code></dt>
+	<dd>Return type of the value set via the <code>@type</code> property or its corresponding javascript typeof value.</dd>
+
+	<dt><code>fn:guessType</code></dt>
+	<dd>
+		<p>The template parser will try to guess the data type of the tested value.</p>
+		<p>This test function will return:</p>
+		<ul>
+			<li><code>rdfs:Container</code> when the value is an array.</li>
+			<li><code>rdfs:Literal</code> for any value that is literal. A literal value is considered to be defined and not an object.</li>
```
[the remark this answers is not in the document]
Comment (duboisp) [reply site/pages/docs/ref/wb-data-json/wb-data-json-en.hbs:448]: fixed the French too

File: site/pages/docs/ref/wb-data-json/wb-data-json-en.hbs [hunk 22]
```diff
@@ -604,6 +659,12 @@
 		<td><code>data-wb-json='{ "url": "", "source": "#idToMyTemplate" }'</code></td>
 		<td>JQuery selector that represent the template element on the current page.</td>
 	</tr>
+	<tr>
+		<td><code>streamline</code></td>
+		<td>Template only. Indicate to not initiate the process by iterating the data and instead execute the mapping object directly</td>
+		<td><code>data-wb-json='{ "url": "", "source": "#idToMyTemplate" }'</code></td>
```
[the remark this answers is not in the document]
Comment (duboisp) [reply site/pages/docs/ref/wb-data-json/wb-data-json-en.hbs:665]: Fixed the French too

File: site/pages/docs/ref/wb-data-json/wb-data-json-en.hbs (hunk 19 above)
Comment (duboisp) [reply site/pages/docs/ref/wb-data-json/wb-data-json-en.hbs:664]: Fixed the French too.

File: site/pages/docs/ref/wb-data-json/wb-data-json-en.hbs (hunk 20 above)
Comment (duboisp) [reply site/pages/docs/ref/wb-data-json/wb-data-json-en.hbs:272]: I added a link to wikipedia about what is an IRI

Comment (duboisp) [reply site/pages/docs/ref/wb-data-json/wb-data-json-en.hbs:272]: And I did the same fix to the French page.

**Review by GormFrank — APPROVED** [review]

**Pull request merged**